spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/MessageConverterConfigurer.java

```diff
@@ -257,105 +259,42 @@ public int selectPartition(Object key, int partitionCount) {
 	 */
-	private final class InboundContentTypeConvertingInterceptor extends ChannelInterceptorAdapter {
-
-		private final MimeType mimeType;
+	private final class InboundContentTypeEnhancingInterceptor extends AbstractContentTypeInterceptor {
 
-		private final CompositeMessageConverterFactory compositeMessageConverterFactory;
-
-		private InboundContentTypeConvertingInterceptor(String contentType, CompositeMessageConverterFactory compositeMessageConverterFactory) {
-			this.mimeType = MessageConverterUtils.getMimeType(contentType);
-			this.compositeMessageConverterFactory = compositeMessageConverterFactory;
+		private InboundContentTypeEnhancingInterceptor(String contentType) {
+			super(contentType);
 		}
 
 		@Override
-		public Message<?> preSend(Message<?> message, MessageChannel channel) {
-			if (message instanceof ErrorMessage) {
-				return message;
-			}
-
-			Message<?> postProcessedMessage = message;
-			MimeType contentType = this.mimeType;
-			if (message.getHeaders().containsKey(MessageHeaders.CONTENT_TYPE)) {
-				Object ct = message.getHeaders().get(MessageHeaders.CONTENT_TYPE);
-				contentType = ct instanceof String ? MimeType.valueOf((String)ct) : (MimeType)ct;
-			}
-
-			boolean deserializationRequired = message.getPayload() instanceof byte[] &&
-					("text".equalsIgnoreCase(contentType.getType()) ||
-					equalTypeAndSubType(MimeTypeUtils.APPLICATION_JSON, contentType) ||
-					equalTypeAndSubType(MessageConverterUtils.X_JAVA_SERIALIZED_OBJECT, contentType) ||
-					equalTypeAndSubType(MessageConverterUtils.X_JAVA_OBJECT, contentType));
-
-			Object payload = deserializationRequired ? this.deserializePayload(message, contentType) : message.getPayload();
-
-			if (payload != null) {
-				Object ct = message.getHeaders().get(BinderHeaders.BINDER_ORIGINAL_CONTENT_TYPE);
-				contentType = ct instanceof String ? MimeType.valueOf((String)ct) : (ct == null ? contentType : (MimeType)ct);
-				postProcessedMessage = MessageConverterConfigurer.this.messageBuilderFactory
-						.withPayload(payload)
-						.copyHeaders(message.getHeaders())
-						.setHeader(MessageHeaders.CONTENT_TYPE, contentType)
-						.removeHeader(BinderHeaders.BINDER_ORIGINAL_CONTENT_TYPE)
-						.build();
+		public Message<?> doPreSend(Message<?> message, MessageChannel channel) {	
+			@SuppressWarnings("unchecked")
+			Map<String, Object> headersMap = (Map<String, Object>) ReflectionUtils.getField(MessageConverterConfigurer.this.headersField, message.getHeaders());
+			
+			/*
+			 * NOTE: The below code for BINDER_ORIGINAL_CONTENT_TYPE is to support legacy message format established 
+			 * in 1.x version of the framework and should/will no longer be supported in 3.x
+			 */
+			Object ct = message.getHeaders().get(BinderHeaders.BINDER_ORIGINAL_CONTENT_TYPE);
+			MimeType contentType = ct instanceof String ? MimeType.valueOf((String)ct) : (ct == null ? this.mimeType : (MimeType)ct);
```

Comment:
Perhaps we should support `byte[]` here too. We would get `byte[]` if the app that generates the message is not a Spring Integration Kafka application (and sets a native `contentType`) Kafka `Header`.

Reply:
Gary, I am not ready to change this since that line is one-for-one with what we had before in MessageSerializer or something. So perhaps a separate issue/test.